### PR TITLE
New C++ Boost Fathead

### LINKED ISSUE #557: C++: Create boost fathead
# Create a new C++ Boost Fathead Instant Answer
Help us make DuckDuckGo the best search engine for programmers!

### What do I need to know?
You'll need to know how to code in **Perl**, **Python**, **Ruby**, or **JavaScript**.
fathead languages

### What am I doing?
You will write a script that scrapes or downloads the data source below, and generates an **output.txt** file containing the parsed documentation. You can learn more about Fatheads and the `output.txt` syntax [**here**](https://docs.duckduckhack.com/resources/fathead-overview.html).

**More specifically**,
Create a fathead that handles boost libraries. This could answer queries like 'cpp boost array`.

**Data source**: http://www.boost.org/doc/libs/1_62_0/

**Instant Answer Page**: *to be created by developer*

### What is the Goal?
As part of our [Programming Mission](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53), we're aiming to reach 100% Instant Answer (IA) coverage for searches related to programming languages by creating new Instant Answers, and improving existing ones.

Here are some Fathead examples:

- C++ reference
  - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/cppreference_doc) | [Example Query](https://duckduckgo.com/?q=cpp+array+cend&ia=about)
- Ruby Docs
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/ruby) | [Example Query](https://duckduckgo.com/?q=array+bsearch&ia=about)
- MDN CSS
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/mdn_css) | [Example Query](https://duckduckgo.com/?q=css+background-position&ia=about)

  css background-position search

[See more related Instant Answers](https://duck.co/ia?repo=fathead)

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a new Instant Answer Page: https://duck.co/ia/new_ia (then let us know, here!)
- [ ] 5) Create the Fathead
- [ ] 6) Create a Pull Request
- [ ] 7) Ping @sahildua2305  for a review

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

### PR DESCRIPTION
**Pull request by nikhilsingh291**
New C++ Boost Fathead

## Description of new Instant Answer, or changes

This fathead gives information about boost libraries for c++.

## Related Issues and Discussions

Fixes #557 
## People to notify

@sahildua2305 

---

Instant Answer Page: https://duck.co/ia/view/cpp_boost

### CODE REVIEW

**Comment by daxtheduck** [comment]
@nikhilsingh291 Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by daxtheduck** [comment]
@nikhilsingh291 Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by daxtheduck** [comment]
@nikhilsingh291 Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/fathead/cpp_boost/fetch.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (moollaza) [inline lib/fathead/cpp_boost/fetch.sh]: Please add `#!/bin/bash`

File: lib/fathead/cpp_boost/parse.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (moollaza) [inline lib/fathead/cpp_boost/parse.sh]: Please add `#!/bin/bash`

**Comment by moollaza** [comment]
@nikhilsingh291 this is a great start. It looks like we need some Fathead triggers here to help make triggering this easier.

Please add a `trigger_words.txt` file alongside the other scripts here. On each line, we should have lowercased words/phrases that we expect to be at the start or end of queries for this Fathead.

E.g.

```
#trigger_words.txt
cpp
cpp boost
c++ boost
```

**Comment by moollaza** [comment]
@nikhilsingh291 it looks like we're just scratching the surface here -- it would be great if we could get more data such as http://www.boost.org/doc/libs/1_62_0/doc/html/align/reference.html#align.reference.align_up

That was a search for "c++ boost align_up` could show the description of the function, the declaration, the requirements and the return values

It looks like there is a "Reference" for each of the function in each of the Boost libraries -- let's get that data too!

It looks like there is also a Matrix page that provides links to more data at well: http://www.boost.org/doc/libs/1_62_0/libs/geometry/doc/html/geometry/matrix.html

Is there any more information you think we can get from here? What parts of the site are most useful to you?

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=uBLAS)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/cpp_boost_mobile.png?4Sr38tHWso)

**Comment by nikhilsingh291** [comment]
@sahildua2305 @moollaza Thanks for the changes requested. Addressed all of them. Let me know about any other changes to be made

**Comment by nikhilsingh291** [comment]
@moollaza Regarding including detailed information of each Boost library. I went through many of the listed libraries, each library seems to have its own way of information organization. Some of them just have examples. 
   Looks like its not possible to include same field of info for all libraries. Please suggest how should I move forward.

**Comment by moollaza** [comment]
> Please suggest how should I move forward.

I think we may need to handle each boost library in an ad-hoc fashion. You may want to look through them individually and collect the pages that we should crawl for more content.

If there is a map or index of all the files on boost.org that may be a starting point. Do you know which of the boost libraries is most popular? We can start with one and see how it goes 😄 

For now this is a good start 👍

**Comment by moollaza** [comment]
@nikhilsingh291 for example, you could check if a `reference.html` page exists for each library, and if so, crawl the reference links in the box at the top.

i.e. http://www.boost.org/doc/libs/1_62_0/doc/html/align/reference.html

**Comment by daxtheduck** [comment]
@nikhilsingh291 this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Pull request merged**